Properly Deal with Scan Speed

Ferociousfeind opened the pull request:
**Feature:** This PR implements the request(s) to more properly replace `(cargo|outfit) scan speed` with a new attribute which behaves differently. You're right, this is a "breaking" change. Original untouched scan speeds will be changed at runtime to be boosted by scan power, so that overall scanning time stays relatively constant at various scan powers, the way scan speed used to work.

## Feature Details
Replaces instances of "scan speed" with "scan efficiency", and catch any remaining instances of "scan speed", say, in the player's save file, dealing with them appropriately.

## UI Screenshots
N/A. Some tooltip adjustments, but just to give scan efficiency its proper tooltip.

## Usage Examples
usage of scan efficiency is unchanged from how scan speed was used, scan speed is used mostly like it was used before the scan change.

Zitchas commented:
Alright, so this replaces Scan Speed with Scan Efficiency?

So Scan Power determines range, and combines somehow with Scan Efficiency and other factors like mass and distance to determine how fast it scans, right?

But you also say that scan speed is used about the same way as it used to be? Does that mean that scan speed is still something? Or is this just referring to the new "scan efficiency"?

Ferociousfeind commented:
Scan speed still exists, but only as a stopgap to catch ships with built-in scan in save files which physically cannot be updated. You _should_ use scan efficiency.

What's happened is in ship.cpp I've replaced "scan speed" with "scan efficiency", and added an extra (two) line(s) that add scan speed too, but multiplied by scan range so that the scanner data from, say, a year ago, would be interpreted by ES as a relatively reasonable scanner outfit (or ship)

So, use scan efficiency, please! Scan speed is deprecated.

Amazinite commented:
If we're deprecating scan speed, then we should do something similar to what was done with the old "cargo scan" and "outfit scan" attributes. That is, warn about them if they exists, and replace them with equivalent values of the new attributes so that we don't need to check for both the new and the old attributes during scanning. (Perhaps scan speed x 16 is what scan efficiency is).

https://github.com/endless-sky/endless-sky/blob/4c5d1d026ffb0925491dff7aa238db20cb2d23ef/source/Outfit.cpp#L323-L344

Ferociousfeind commented:
That's a good point......
It'd be better to modify outfit.cpp
I'll, uh, do that.

Ferociousfeind commented:
I am back from vacation. The integration tests are failing. I believe this is because "scan speed" gets set to 0, which makes the Heron look like it's erroneously incapable of scanning, when it is the integration test that is wrong.

I don't know what to do about this.

Hurleveur commented:
At the very least the test for flagship attribute now needs to check for scan efficiency,thats why it fajls

RisingLeaf reviewed (changes requested):
Data file changes + code file changes.
I do not like this. I feel we should have backward support, especially because of plugins.
Then do data file changes in a follow up.

Ferociousfeind commented:
Why must the data file changes be in a followup PR? In what situation do the data file changes get in but the underlying code changes do not? Or vice versa?

RisingLeaf commented:
> Why must the data file changes be in a followup PR? In what situation do the data file changes get in but the underlying code changes do not? Or vice versa?

we separate code from data files, ~~deal with it~~

warp-core commented:
> I feel we should have backward support, especially because of plugins.

As I understand it, the purpose of this PR is to ensure that plugins (and ships in save files with the old attributes) don't have ridiculously long scan times.

> we separate code from data files, deal with it

Only when it makes sense. It doesn't make sense here.
In fact, this PR introduces warnings that I think will cause the tests to fail if the accompanying data changes aren't made.
I can think of two of my own PRs relatively recently that made changes to both the engine code and data: #7413 and 
7600. The latter of which is somewhat similar to this one.

Hurleveur commented:
Yeah this PR should have data changes with it, and code changes are relatively small anyway

Zitchas commented on data/human/outfits.txt:
```diff
@@ -251,9 +251,9 @@ outfit "Surveillance Pod"
 	"outfit scan power" 16
-	"outfit scan speed" 2
+	"outfit scan efficiency" 20
 	"cargo scan power" 4
-	"cargo scan speed" 2
+	"cargo scan efficiency" 10
```
If I'm reading this correctly, this is buffing the Surveillance pod, correct? 

Just noting that everywhere else translates the exact same value from scan speed directly into scan efficiency, but the surveillance pod and the Kas'ik Tek 7 are the two exceptions where the value has been increased.

Ferociousfeind replied:
Most outfits had `"(cargo|outfit) scan speed" 1`, which meant a 60 frame scan time. (A scan speed of 2 equals about 43 frames, AKA 60 / sqrt(2) frames.) A decent translation is making scan efficiency to be somewhere around 15x original value, or set efficiency to speed * power.

So, this isn't as much a buff as it is "accounting for inflation" because the scan speed was 2 rather than 1, it warrants a higher scan efficiency

Zitchas commented on data/korath/korath ships.txt:
```diff
@@ -404,9 +404,9 @@ ship "Kas'ik Tek 7"
 		"outfit scan power" 18
-		"outfit scan speed" 2
+		"outfit scan efficiency" 20
 		"cargo scan power" 6
-		"cargo scan speed" 2
+		"cargo scan efficiency" 10
```
Is there a reason that outfit scan speed->efficiency gets a 10x multiplier, but the cargo one gets 5x?

Ferociousfeind replied:
for most ships, cargo = smol, outfit = beeg, and that's about it.

Ferociousfeind replied:
(Also cargo scan power = 1/3 outfit scan power on the kas'ik tek, so it seemed like it was prioritizing outfits over cargo)

Zitchas reviewed (commented):
I can't comment on the code.

Three questions:
1) Based on the tips, it looks like scan speed isn't just the deprecated name for efficiency, but something that could actually be used to boost the scan speed more directly (aka, without being chopped down by the square root). Is this the case? Or are the two actually synonymous?
2) What happens if someone makes an outfit that has both a scan speed AND scan efficiency?
3) Most data entries are just the same value copied over. The exception being the surveillance scanner and one Korath ship. In these two cases, the outfit scan efficiency is 10x, and the cargo scan efficiency is 5x the original value. What's the reason for this?

Aside from those points, the data all looks like it is directly transferred over. I do feel that with the new mechanic the overall values for scanning need a rework to verify that the scanning "mini game" is working as intended everywhere, but that's not something that should be in this PR. This is just the baseline and making sure everything mechanically works properly, and as far as that goes, with the exception of the values for those two places I asked about above, this looks good.

I'd like to see this merged sooner rather than later so that I can work on the Remnant aspect of this.

Hurleveur commented:
Scan efficiency is supposed to replace scan speed that is not to be used again.
So both should not be present on the same outfit.

Zitchas commented:
But if it is present? Is that an error the game should check for and handle, or will it do something else?

For the sake of argument, say a plug-in has both on one scanner.

Hurleveur commented:
Both will be added on top of each others with the speed being considered as efficiency/15

Ferociousfeind commented:
If @Zitchas intends to tune up the Remnant built-in scanning values, that'd be awesome, because there are a lot of them, and I think I did a poor job because I didn't properly consider what the values _meant._ As I said, a single outfit scanner is 2 tons and that's not enough (in my opinion) to get an OK scanning time. So, when I transcribed the remnant stuffs, I don't think I've sufficiently increased the scanning efficiency for these ships to be OK standalone scanners.

But also IDK, I didn't test each ship individually, they _might_ be fine.

(still psyched over my scanner changes getting in, I'm just also sad that my _negligence_ broke the scanning game)

Zitchas reviewed (approved):
I can't comment on the code part, but the values in the data look good.

As a general thought, I think scanners may be one area where we'll need to pull back from comparing them faction to faction, and spending more time looking at them in terms of "how are they suppose to handle the scanning minigame in their region/ecosystem?" and rebalance from there based on how important scanning is to them, how much scanning is done in their area, and what they typically have to scan.

I think the data values as presented here are good, and should be merged sooner rather than later to get scanning working across the board. But I also think that the faction leads for each region should be planning on taking a hard look at these values and potentially re-balancing them accordingly. 

And yes, I'd be happy to do so for the Remnant and the Ember Waste factions.

Hurleveur reviewed (approved):
Code wise this is very simple and looks nice.
Data wise Grey and I already looked at Hai stuff (ladybug, value detector) it should be good as is.
Worth noting this will divide by 2 the importance of distance which makes sense because it was too high.

This alsl fixes any plugins that were using scan speed

warp-core reviewed (commented):
Not something changed here, but I notice the `doScan` lambda just has a `&` in the captures? Maybe just capture the needed variables?

warp-core commented:
Something I noticed, existing ships in saves from after the previous scan changes are going to get ridiculous scan efficiencies now. Kind of amusing.

Amazinite commented on data/tooltips.txt:
```diff
@@ -81,6 +81,9 @@ tip "cargo scan power:"
 	`The range of this cargo scanner is proportional to the square root of this value.`
 
 tip "cargo scan speed:"
+	`A higher scan speed will correspond to a quicker scan.`
```
I'd say we should revert these tooltips to their old versions before the scan speed changes, but the old tooltips aren't true anymore since we're converting to efficiency.
> Determines how long it takes to perform a cargo scan. A total value of 1 means one second, and the time decreases in proportion to the square root of this value, e.g. 4 means half a second.

I think at least taking the first sentence would be better, though.
```suggestion
	`Determines how long it takes to perform a cargo scan. A higher value will correspond to a quicker scan.`
```

Amazinite commented on data/tooltips.txt:
```diff
@@ -375,6 +378,9 @@ tip "outfit scan power:"
 	`The range of this outfit scanner is proportional to the square root of this value.`
 
 tip "outfit scan speed:"
+	`A higher scan speed will correspond to a quicker scan.`
```
```suggestion
	``Determines how long it takes to perform an outfit scan. A higher value will correspond to a quicker scan.`
```

Amazinite commented on source/Outfit.cpp:
```diff
@@ -341,8 +341,23 @@ void Outfit::Load(const DataNode &node)
+		}
+
+		// Similar check for scan speed which is replaced with scan efficiency.
+		// If this could be simplified into the above, I wouldn't know how to.
```
```suggestion
```

Amazinite commented on source/Outfit.cpp:
```diff
@@ -341,8 +341,23 @@ void Outfit::Load(const DataNode &node)
+			attributes[label] = 0.;
+			node.PrintTrace("Warning: Deprecated use of \"" + label + "\" instead of \""
+					+ kind + " scan efficiency\":");
+			// A reasonable update is 15x the previous value-
+			// scans take 10x longer standard, large ships take longer to scan,
+			// and scanning speed scales inversely with distance.
```
```suggestion
			// A reasonable update is 15x the previous value, as the base scan time
			// is 10x what it was before scan efficiency was introduced, along with
			// ships which are larger or further away also increasing the scan time.
```

Amazinite commented on source/Ship.cpp:
```diff
@@ -2665,16 +2665,18 @@ int Ship::Scan(const PlayerInfo &player)
 	// A ship with less than 10 tons of outfit space or cargo space takes as long to
-	// scan as one with 10 tons. This avoids small sizes being scanned instantly, or
-	// causing a divide by zero error at sizes of 0.
+	// scan as one with 10 tons. This avoids small sizes being scanned instantly.
+	// If instantly scanning very small ships is desirable, this can be removed.
```
I mean, the divide by zero comment is still true. A depth of 0 would result in a denominator of 0. I'd say better to keep the mention of dividing by zero so that if we do remove this lower limit, the comments at least call out that we need to watch for this.
```suggestion
	// scan as one with 10 tons. This avoids small sizes being scanned instantly, or
	// causing a divide by zero error at sizes of 0.
	// If instantly scanning very small ships is desirable, this can be removed.
```

Amazinite reviewed (commented):
I feel like there was something else I wanted to double check with this swap out, but I forget what it was. Must not be important.
